### PR TITLE
Add initial regression script

### LINKED ISSUE #2718: Optional constructor arguments are lost
Issue created from [fantomas-online](https://fsprojects.github.io/fantomas-tools/#/fantomas/main?data=N4KABGBEDGD2AmBTSAuKAXAngB0WAwgM4BuAYgJYA2iAOgE40B2EE2d5xAhurQ8ywAp6TFqLB1EnJHVIBXRtDRyFAHgAqiAB7oASpOkA%2BADTD%2BYsAG0VAeWzpysRp0oGAumAD8hRNk51usHSEJnzmEFa29o7Obp4AjrKwPCEiYRF2Dk4u7h4AFpyEABL6iEEpZmLpUVmxHuQA5oyBiACidHSBwaZhljYZ0dmehADW5Ng6sADuhN0sAJRgBWDoueSEYAC8sxDkjLml5OgEJBTUKkTEE5MG22JCoT0sytAqAPpGYO%2BfBgIAZvLLVbrOCUWQAW0Y6wAtAZjpcpgIVmswCgPLCLqdEB8QeDIXM5uVHk95C8vq8fv9mB1JmAYeIpgA6fCwUEQwgE249CRSUrPQlEsBIX6cWSUdAAQTo9TA3l8-nQnTANEgyv5RKFIrFkulCSSeAA5Mr9WrHhrRRKpWB8kUSkFlnRZFjOWEzVrLQ0mhI2h07cLKN4TT1XRbpSMxld1gAGZ0QOaQIzgKDedD2Rj1QioSyJiCgcyQAAkWFwmcgu3Q8ezLALXFByDQFkrokjCZ6kAAkowkIx0ABlcgAL2QLbCABZG2BXJWAL7DnPjgtFutQMsVvP5muOzMNnoARlnYkgAFlOJoADK7RCnxBplarsI7gBMkfHk9EM8ruYPhZwS8g1-g1i-OejCID2WDUHeojVs4m71uOEAPvuUEtJ2gHAUO8FQJQvyQC%2B077p%2BUHfsWaCQAARrALKQVW64wUu25hAAzEhVYdt4dDoBQWQAHKIJMlAXtRojoA6iB4W%2BBHzsRv4UVRLHQbWW6YSOLEQJAPa%2BNAiAAEKIL8zQAAp%2BJwYKIDwdBCSwImOuJLDvqIhE0YuJayZQlkKbBWY9AArKpUAaZwWm6fpEinlMpTQAUiAdsQsCRdUlkQNZYnmK%2BdmSWuzmka57m0YpcE9AAbH56maTpenNAAqtguB0JF3gxXF3CZIlYB%2Bt4tkQPZLCOWp0kuZRbnyXlnkMeYADsJUBUFFUSPglAFIQzKQtZ0AKhZfntSlYhpV1GVfllUA5cNG70ZhAAcU1lcFzSHogYJkaUrVbZ1YDdXOmU-gNclSadSk9AAnFdgXlSFiDMpQjjPc4HWpfhH5SYd5GDblf0FfezaYaVIPir85nMmCYKcK1yWve9YC9VA-XZSjJ10f9957lj02g80Pb3eQIJQ5tMPbaIu1vftRFI8dv30%2Bj5iPsDWm4%2BZ7NgpzLLc5hpNwxJCOfSRR202L%2BVefezHM2VkqwPI8AACKIAJCvmSTolk0LTlfaRK503rY1iDuKlY8emhtr8aj7IwPa5IE6AAOrkPAt5%2Bc%2BavpRrB3O8u3ao%2BL%2BuS75Psnv7gfXi0-qgaHHGR9HuStYVcc7fDDmI8npap27o2YTuxXZ37jC-OQmi2KUAR0C0mhsIghCEC1fnnVX-M1z1ddaw35ZN2du6Te3ehwHQ8ClzHylTywAvk5TC7167uvN7ul1r4gG-wNx4KPXQgFtjwYIZn5O4O4nwv12C5pUBepBAhExTKUNQX0l4M0lkDLG69AjwEPH-ASIFAF0GAbbEq0B8j%2BDWqUV4kwo63k-rXTWv5T5rjRhnMQT4Sq%2B0lP4TA1g6DnkIBHAh5cJ57wgAfR2fUkZkK-BQj2ogHxM1bLQ9onAGFMLWOgO%2BD1ShPxfm-FuRDZ4kJLL-MU-9kFAO4OZMBxYIESyoYhLGdDJGMOYegBBWikGIBQWgp6GCsGBXMngthuF457S-k7ee-CiKCMwg%2BQ2YiTwADU6LaV2PAXY9Rt7sIupwicM8PpJz8Y3M%2By8wgPm9qEzQzxqhRM7LE%2BJrURxJO4T43hJ8MnkPTkIlgD4s55PNkkAA4mZQew9R6ZFKRw1RqTv7z1FnU92QS26tl4vxC8ul0CTEQNeAxiBLZd0YIcTI4pOx3XkUEO2NkvGCyqVTEWOtRnn2yavVs4oBKNAKZkPsjRuCyAkGoWAHYuzoGasrHoL0DmHznjJU5Aj6lBMvlcsUpQnD2GIJeRw9RtkPxWbsdZjhlE-N5gMimALvpDUyZAqh0CxF-3QqeEyZF4CcHmrAMeaYpl2OhoXTFR9qbax%2BmcrJ5hGKY1bF0g4pluzOAANILOwO8686AOzaX8AoBJ6LGV-J4cc%2BuIzgVjJ6IxURYRIDaQWowYY6FCAmzNrxFhiB4GIIvF0iQPTUV7L5vvFJWL1E0zZaq85nLTGtm0n4G6EhzZrGgOwBWUKzWVTWY4S20AFrynHphX51d1bELSb%2BTR9g7FSsCsMMyYFMAQSMZQ0QjEQlapsWmmZ2Cs29nAhhVsgaTK4HgJ4hNCck1DNIbUt1HKxCMVyVq4VPhfZyIfoBHVnA9UGtajuSMFTHXMpOa6gJIL1XNK1WBdga1DwIBrWEeN083yJlcMOUshBSBj0zC9KcQA)

#### Code

```fsharp
type CsvFile
    private
    (
        readerFunc: Func<TextReader>,
        [<Optional>] ?separators,
        [<Optional>] ?quote,
        [<Optional>] ?hasHeaders,
        [<Optional>] ?ignoreErrors,
        [<Optional>] ?skipRows
    ) as this =
    inherit CsvFile<CsvRow>
        (
            Func<_, _, _>(fun this columns -> CsvRow(this :?> CsvFile, columns)),
            Func<_, _>(fun row -> row.Columns),
            readerFunc,
            defaultArg separators "",
            defaultArg quote '"',
            defaultArg hasHeaders true,
            defaultArg ignoreErrors false,
            defaultArg skipRows 0
        )
```
            

#### Result

```fsharp
type CsvFile private (readerFunc: Func<TextReader>) as this =
    inherit
        CsvFile<CsvRow>(
            Func<_, _, _>(fun this columns -> CsvRow(this :?> CsvFile, columns)),
            Func<_, _>(fun row -> row.Columns),
            readerFunc,
            defaultArg separators "",
            defaultArg quote '"',
            defaultArg hasHeaders true,
            defaultArg ignoreErrors false,
            defaultArg skipRows 0
        )

```
            
#### Problem description

Most likely a problem in ASTTransformat when constructing simplePats.

#### Extra information

- [x] The formatted result breaks my code.
- [ ] The formatted result gives compiler warnings.
- [ ] I or my company would be willing to help fix this.

#### Options

Fantomas main branch at 2023-01-14T10:13:32Z - 5336c50d3aee76b59c41af2b17926ac4e1557d8c

Default Fantomas configuration

<sub>Did you know that you can ignore files when formatting from fantomas-tool or the FAKE targets by using a [.fantomasignore file](https://fsprojects.github.io/fantomas/docs/end-users/IgnoreFiles.html)?</sub>

### PR DESCRIPTION
**Pull request by nojaf**
This is the first attempt to introduce some more regression testing.
The idea is to clone a repository (that already uses Fantomas), build it, and format it using the latest code. Afterwards, build the repository again to verify that still works.

This isn't a silver bullet and does not detect absolutely everything, but I think it is a start. 
We could run this manually before we are going to release a new version. I don't think we want this before every PR.

All feedback is welcome.

Todo:
- [x] Hook into GitHub Actions
- [x] Fix #2718

### CODE REVIEW

File: tests/regressions.fsx [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+        printfn $"Completed run for {repository.Name}"
+    }
+
+// Ensure Fantomas dll is built
+(wrap fantomasRoot "dotnet" $"build -c Release {fantomasProject}").Wait()
+
+// Run repositories
+for repo in repositories do
+    (run repo).Wait()
```
Comment (nojaf) [inline tests/regressions.fsx:101]: This `Wait` bit might be sacrilege, please let me know what the idiomatic thing to do here in F# is.

Comment (dawedawe) [reply tests/regressions.fsx:101]: How about
```fsharp
let t = run repo
t |> Async.AwaitTask |> Async.RunSynchronously
```

**Comment by dawedawe** [comment]
I like it a lot and yes, running this manually before a release or whenever one feels like it should be enough as it's a rather big work load.

File: tests/regressions.fsx (hunk 1 above)
Comment (nojaf) [reply tests/regressions.fsx:101]: I thought of that but it feels a bit strange to move to the `Async` world at the end for this.
As the tasks are hot, I think staying in the `Task` world makes more sense.

**Comment by nojaf** [comment]
I'm going to merge this to be able to try out the GitHub Action, we can refine the script later on.

**Pull request merged**